src/Microsoft.ML.TimeSeries/SrCnnAnomalyDetectionBase.cs

```diff
@@ -0,0 +1,298 @@
+        bool ITransformer.IsRowToRowMapper => ((ITransformer)InternalTransform).IsRowToRowMapper;
+
+        /// <summary>
+        /// Create a clone of the transformer. Used for taking the snapshot of the state.
+        /// </summary>
+        IStatefulTransformer IStatefulTransformer.Clone() => InternalTransform.Clone();
+
+        /// <summary>
+        /// Schema propagation for transformers.
+        /// Returns the output schema of the data, if the input schema is like the one provided.
+        /// </summary>
+        public DataViewSchema GetOutputSchema(DataViewSchema inputSchema) => InternalTransform.GetOutputSchema(inputSchema);
+
+        /// <summary>
+        /// Constructs a row-to-row mapper based on an input schema. If <see cref="ITransformer.IsRowToRowMapper"/>
+        /// is <c>false</c>, then an exception should be thrown. If the input schema is in any way
+        /// unsuitable for constructing the mapper, an exception should likewise be thrown.
+        /// </summary>
+        /// <param name="inputSchema">The input schema for which we should get the mapper.</param>
+        /// <returns>The row to row mapper.</returns>
+        IRowToRowMapper ITransformer.GetRowToRowMapper(DataViewSchema inputSchema)
+            => ((ITransformer)InternalTransform).GetRowToRowMapper(inputSchema);
+
+        /// <summary>
+        /// Same as <see cref="ITransformer.GetRowToRowMapper(DataViewSchema)"/> but also supports mechanism to save the state.
+        /// </summary>
+        /// <param name="inputSchema">The input schema for which we should get the mapper.</param>
+        /// <returns>The row to row mapper.</returns>
+        public IRowToRowMapper GetStatefulRowToRowMapper(DataViewSchema inputSchema)
+            => ((IStatefulTransformer)InternalTransform).GetStatefulRowToRowMapper(inputSchema);
+
+        /// <summary>
+        /// Initialize a transformer which will do lambda transfrom on input data in prediction engine. No actual transformations happen here, just schema validation.
+        /// </summary>
+        public IDataView Transform(IDataView input) => InternalTransform.Transform(input);
+
+        /// <summary>
+        /// For saving a model into a repository.
+        /// </summary>
+        void ICanSaveModel.Save(ModelSaveContext ctx) => SaveModel(ctx);
+
+        private protected virtual void SaveModel(ModelSaveContext ctx)
+        {
+            InternalTransform.SaveThis(ctx);
+        }
+
+        /// <summary>
+        /// Creates a row mapper from Schema.
+        /// </summary>
+        internal IStatefulRowMapper MakeRowMapper(DataViewSchema schema) => InternalTransform.MakeRowMapper(schema);
+
+        /// <summary>
+        /// Creates an IDataTransform from an IDataView.
+        /// </summary>
+        internal IDataTransform MakeDataTransform(IDataView input) => InternalTransform.MakeDataTransform(input);
+
+        internal SrCnnAnomalyDetectionBaseCore InternalTransform;
```

Comment:
Is this one `readonly`? Or it contains mutable states?

Reply:
It's readonly.